# Проект: Рынок видеоигр

- Автор: Бураевская Александра
- Дата: 03.04.2026

### Цели и задачи проекта

Используем данные датасета /datasets/new_games.csv, который содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр. Познакомимся с этими данными, проверим их корректность и проведем предобработку, получив необходимый срез данных.

### Описание данных

В проекте будут использованы данные датасета new_games.csv с таким описанием:
- `Name` — название игры;
- `Platform` — название платформы;
- `Year of Release` — год выпуска игры;
- `Genre` — жанр игры;
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий);
- `EU sales` — продажи в Европе (в миллионах проданных копий);
- `JP sales` — продажи в Японии (в миллионах проданных копий);
- `Other sales` — продажи в других странах (в миллионах проданных копий);
- `Critic Score` — оценка критиков (от 0 до 100);
- `User Score` — оценка пользователей (от 0 до 10);
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта

1. Загрузка и знакомство с данными
2. Проверка ошибок в данных и их предобработка
3. Фильтрация данных
4. Категоризация данных
---

## Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
# Импортируем библиотеку pandas
import pandas as pd

In [2]:
# Выгружаем данные
new_games = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [3]:
# Выводим информацию о датафрейме
new_games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
# Выводим первые строки датафрейма на экран
new_games.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

Датасет `new_games.csv` содержит 11 столбцов и 16956 строк.

Данные датасета `new_games.csv` соответствуют описанию.

В датасете `new_games.csv` присутствуют пропуски в следующих столбцах: `Name`, `Year of Release`, `Genre`, `Critic Score`, `User Score` и `Rating`.

Изучим типы данных и их корректность:
- **Числовые значения с плавающей запятой (float64).** Четыре столбца имеют тип данных `float64`:
    - `Year of Release` - год выпуска игры должен быть целочисленным значением (int64),
    - `NA sales` и `Other sales` - это продажи в миллионах проданных копий, так что тип данных верный,
    - `Critic Score` - это оценка критиков (от 0 до 100), скорее всего здесь должны быть целочисленные значения (int64).
- **Строковые данные (object).** Восемь столбцов имеют тип данных `object`:
    - `Name`, `Platform` и `Genre` содержат строковую информацию (название игры, платформы и жанр игры), что подходит для текстовых данных. Здесь тип данных указан корректно,
    - `EU sales`, `JP sales` и `User Score` - здесь представлены данные о продажах в миллионах проданных копий и оценка пользователей (от 0 до 10), которым больше подходит тип данных `float64`,
    - `Rating`также хранят текстовые данные, но их можно рассматривать как категориальные признаки. и можно использовать тип `category`.

После анализа типов данных видно, что 5 столбцов из 11 некорректно представлены. 

В датасете `new_games.csv` названия столбцов отражают содержимое данных и прописаны в удобном для работы виде.

---

## Проверка ошибок в данных и их предобработка


### Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [5]:
total_row = new_games.shape[0]
total_row

16956

In [6]:
# Выводим названия столбцов датафрейма
new_games.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [7]:
# Сохраняем в список желаемые названия столбцов
snake_case_lower_columns = [
    'name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales', 'jp_sales', 
    'other_sales', 'critic_score', 'user_score', 'rating']

# Передаём список атрибуту columns датафрейма
new_games.columns = snake_case_lower_columns

# Выводим скорректированные названия столбцов датафрейма
new_games.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

### Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [8]:
# приводим столбцы 'eu_sales', 'jp_sales' без пропусков типа данных object к числовому виду
new_games['eu_sales'] = pd.to_numeric(new_games['eu_sales'], errors='coerce')
new_games['jp_sales'] = pd.to_numeric(new_games['jp_sales'], errors='coerce')
new_games['user_score'] = pd.to_numeric(new_games['user_score'], errors='coerce')

In [9]:
# Проверяем типы данных
new_games.dtypes

name                object
platform            object
year_of_release    float64
genre               object
na_sales           float64
eu_sales           float64
jp_sales           float64
other_sales        float64
critic_score       float64
user_score         float64
rating              object
dtype: object

### Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [10]:
# Выводим количество пропущенных строк в датафрейме

missing_count = new_games.isna().sum()
missing_share = new_games.isna().mean() * 100

missing = pd.DataFrame({
    'missing_count' : missing_count,
    'missing_share' : missing_share
})

missing.sort_values(by='missing_count', ascending=False)

,missing_count,missing_share
user_score,9268,54.659118
critic_score,8714,51.391838
rating,6871,40.522529
year_of_release,275,1.621845
eu_sales,6,0.035386
jp_sales,4,0.023590
name,2,0.011795
genre,2,0.011795
platform,0,0.000000
na_sales,0,0.000000


- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


Пропуски обнаружились в 8 столбцах:
- `name` - 2 пропуска
- `year_of_release` - 275 пропуска
- `genre` - 2 пропуска
- `eu_sales` - 6 пропуска
- `jp_sales` - 4 пропуска
- `critic_score` - 8714 пропуска
- `user_score` - 9268 пропуска
- `rating` - 6871 пропуска

Самая большая доля пропусков в столбцах `critic_score`, `user_score` и `rating`, и это может исказить результат. 

В столбцах `critic_score`, `user_score` и `rating` 0 не подходит, т.к. это может означать 0 очков и 0й рейтинг, поэтому лучше взять -1. 

Строки с пропусками в столбцах `name` и `year_of_release` можно удалить, т.к. такие данные не подходят для рещения задач проекта.

Пропуски в столбце `genre` можно проигнорировать, т.к. данные в этом столбце не важны для анализа.

В столбцах `eu_sales` и `jp_sales` пропуски можно заменить на среднее значение.


- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [11]:
# Отбираем строки без пропусков в столбце date
new_games = new_games.dropna(subset=['name', 'year_of_release'])

In [12]:
# Заменяем пропуски
new_games['critic_score'] = new_games['critic_score'].fillna(-1)
new_games['user_score'] = new_games['user_score'].fillna(-1)
new_games['rating'] = new_games['rating'].fillna("-1")

In [13]:
# проверим получилось ли убрать пропуски
new_games.isna().sum()

name               0
platform           0
year_of_release    0
genre              0
na_sales           0
eu_sales           6
jp_sales           4
other_sales        0
critic_score       0
user_score         0
rating             0
dtype: int64

In [14]:
# приводим столбцы 'year_of_release' и 'critic_score' к типу int64
new_games[['year_of_release', 'critic_score']] = new_games[['year_of_release', 'critic_score']].astype('int64')
# приводим столбцы user_score' к типу float64
new_games['user_score'] = new_games['user_score'].astype('float64')

In [15]:
# Проверяем типы данных
new_games.dtypes

name                object
platform            object
year_of_release      int64
genre               object
na_sales           float64
eu_sales           float64
jp_sales           float64
other_sales        float64
critic_score         int64
user_score         float64
rating              object
dtype: object

### Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [16]:
# выводим уникальные значения
for col in ['name', 'platform', 'genre', 'rating']:
    print(new_games[col].unique())

['Wii Sports' 'Super Mario Bros.' 'Mario Kart Wii' ...
 'Woody Woodpecker in Crazy Castle 5' 'LMA Manager 2007'
 'Haitaka no Psychedelica']
['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'WiiU' 'GC' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX']
['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']
['E' '-1' 'M' 'T' 'E10+' 'K-A' 'AO' 'EC' 'RP']


In [17]:
# подсчитаем количество уникальных значений
for col in ['name', 'platform', 'genre', 'rating']:
    print(new_games[col].nunique(),col)

11426 name
31 platform
24 genre
9 rating


In [18]:
# удаляем пробелы в начале и конце
for col in ['name', 'platform', 'genre', 'rating']:
    new_games[col] = new_games[col].str.strip()

In [19]:
# приводим к верхнему регистру
for col in ['platform', 'genre', 'rating']:
    new_games[col] = new_games[col].str.upper()

In [20]:
# подсчитаем количество уникальных значений
for col in ['name', 'platform', 'genre', 'rating']:
    print(new_games[col].nunique(),col)

11426 name
31 platform
12 genre
9 rating


Видно, что количество уникальных жанров снизилось на половину. Значит один жарн был написан разными регистрами.

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [21]:
# Считаем количество уникальных значений в столбцах 'name','year_of_release','platform','genre'
print("Количество явных дубликатов: ", new_games[['name','year_of_release','platform','genre']].duplicated().sum())

Количество явных дубликатов:  236


In [22]:
# Определите количество строк в датафрейме до удаления дубликатов
initial_row_count = len(new_games)
print(f'Количество строк до удаления дубликатов: {initial_row_count}')

Количество строк до удаления дубликатов: 16679


In [23]:
# Удалите дублирующиеся строки
new_games = new_games.drop_duplicates(subset=['name', 'platform', 'year_of_release', 'genre'])

In [24]:
final_row_count = new_games.shape[0]
print(f"Количество строк после удаления дубликатов: {final_row_count}")

Количество строк после удаления дубликатов: 16443


- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

Количество найденый явных дубликатов 236

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [25]:
print("Всего было удалено ",total_row - final_row_count," сторк,", round((1-final_row_count/total_row)*100), "% от изначального количества строк")

Всего было удалено  513  сторк, 3 % от изначального количества строк


- После проведения предобработки данных напишите общий промежуточный вывод.

- Проведена проверка ошибок в данных и их предобработка. 
- Удалены строки с пропусками в столбцах 'name' и 'year_of_release'. 
- Заменили пропуски на значение -1 в столбцах 'critic_score', 'user_score', 'rating'
- Привели названия фильмов к нижнему регистру, данные в столбцах platform', 'genre', 'rating' приводим к верхнему регистру и 
- Удалили лишние пробелы в столбцах 'platform', 'genre', 'rating'.
- Удалили 236 явных дубликатов.

---

## Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [26]:
# фильтруем данные
df_actual = new_games[(new_games['year_of_release'] >= 2000) & (new_games['year_of_release'] <= 2013)].copy()

In [27]:
# подсчитываем количество строк
df_actual.shape[0]

12780

---

## Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [28]:
# Разбиваем на категории с помощью
def categorize_user_scores(score):
    if 8 <= score <= 10:
        return "высокая оценка"
    elif 3 <= score < 8:
        return "средняя оценка"
    elif 0 <= score < 3:
        return "низкая оценка"
    else:
        return "без оценки" 
    
df_actual['category_user_score'] = df_actual['user_score'].apply(categorize_user_scores)

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [29]:
# Разбиваем на категории с помощью pd.cut
def categorize_critic_scores(score):
    if 80 <= score <= 100:
        return "высокая оценка"
    elif 30 <= score < 80:
        return "средняя оценка"
    elif 0 <= score < 30:
        return "низкая оценка"
    else:
        return "без оценки" 
    
df_actual['category_critic_score'] = df_actual['critic_score'].apply(categorize_critic_scores)

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [30]:
# группируем
df_actual.groupby(['category_user_score'])['name'].count()

category_user_score
без оценки        6298
высокая оценка    2286
низкая оценка      116
средняя оценка    4080
Name: name, dtype: int64

In [31]:
# группируем
df_actual.groupby(['category_critic_score'])['name'].count()

category_critic_score
без оценки        5612
высокая оценка    1691
низкая оценка       55
средняя оценка    5422
Name: name, dtype: int64

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [32]:
# подсчитаем количество уникальных значений 
df_actual['platform'].value_counts().head(7)

PS2     2127
DS      2120
WII     1275
PSP     1180
X360    1121
PS3     1086
GBA      811
Name: platform, dtype: int64

---

## Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

- Сначала познакомились с данными: прочитали описание, посмотрели объем данных, и проверили наличие пропусков.
- Была выполнена предобработка данных: исправили тип данных, заполнили пропуски, удалили дубликаты.
- Оставили только актуальные данные для анализа - игры, выпущенные с 2000 по 2023 год.
- Добавили 2 столбца с категориями оценок пользователей и критиков.
- Выделили топ-7 платформ по количеству выпущенных игр, а именно - PS2, DS, WII, PSP, X360, PS3, GBA.
- Около 50 % игр не имеют оценок как и от пользователей, так и от критиков. Это может говорить о низкой популярности среди пользователей и незаинтересованности критиков. Кроме этого, возможно произошла ошибка при вводе данных. 